format/tap/tap.go

```diff
@@ -0,0 +1,139 @@
+package tzx
+
+// https://worldofspectrum.net/zx-modules/fileformats/tapformat.html
+
+import (
+	"embed"
+
+	"golang.org/x/text/encoding/charmap"
+
+	"github.com/wader/fq/format"
+	"github.com/wader/fq/pkg/decode"
+	"github.com/wader/fq/pkg/interp"
+	"github.com/wader/fq/pkg/scalar"
+)
+
+//go:embed tap.md
+var tapFS embed.FS
+
+func init() {
+	interp.RegisterFormat(
+		format.TAP,
+		&decode.Format{
+			Description: "TAP tape format for ZX Spectrum computers",
+			DecodeFn:    tapDecoder,
+		})
+
+	interp.RegisterFS(tapFS)
+}
+
+// The TAP- (and BLK-) format is nearly a direct copy of the data that is stored
+// in real tapes, as it is written by the ROM save routine of the ZX-Spectrum.
+// A TAP file is simply one data block or a group of 2 or more data blocks, one
+// followed after the other. The TAP file may be empty.
+func tapDecoder(d *decode.D) any {
+	d.Endian = decode.LittleEndian
+
+	d.FieldArray("blocks", func(d *decode.D) {
+		for !d.End() {
+			d.FieldStruct("block", func(d *decode.D) {
+				decodeTapBlock(d)
+			})
+		}
+	})
+	return nil
+}
+
+func decodeTapBlock(d *decode.D) {
+	// Length of the following data.
+	length := d.FieldU16("length")
+
+	// read header, fragment, or data block
+	switch length {
+	case 0:
+		// fragment with no data
+	case 1:
+		d.FieldRawLen("data", 8)
```

Comment:
Possible future change: Are these data fragments (and data below) something that can be joined into something decodable? if so fq can append things together and decode a "sub"-buffer of some format or just provide a large raw field